src/qw/md.py

```diff
@@ -28,3 +31,292 @@ def text_under_heading(text: str, heading: str) -> str:
         raise QwError(msg)
 
     return "\n".join(sub_heading_lines).strip()
+
+
+# Breaking Markdown text into paragraphs of some type
+FENCED_RE = re.compile(r"(?:^|\n+)(?:~{3,}|`{3,})\s*(.*?)\n([\S\s]*?)\n(?:~{3,}|`{3,})(?:$|\n+)")
+PARAGRAPH_RE = re.compile(r"\n{2,}|\n(?=\s*[\*\+\-]\s)")
+
+# Determinig the type of a paragraph
+## - item
+LISTU_RE = re.compile(r"(\s*)[\*\+\-]\s+(.*)")
+## 1. item
+LISTO_RE = re.compile(r"(\s*)\d+\.\s+(.*)")
+
+# Breaking Markdown paragraphs into runs of some type
+## **bold**
+ASTERISK2_RE = re.compile(r"\*\*(\S(?:.*\S)?)\*\*(?=[^*])")
+## __bold__
+UNDERSCORE2_RE = re.compile(r"(?:^|(?<=\s))__(\S(?:.*\S)?)__(?:$|(?=\s))")
+## *italic*
+ASTERISK1_RE = re.compile(r"\*(\S(?:.*\S)?)\*")
+## _italic_
+UNDERSCORE1_RE = re.compile(r"(?:^|(?<=\s))_(\S(?:.*\S)?)_(?:$|(?=\s))")
+## `preformated`
+BACKTICK_RE = re.compile(r"`([^` \t\r\n].*?)`")
+## [[link]]
+SQUARE2_RE = re.compile(r"\[\[(.+?)\]\](?=[^\]])")
+## [text](link)
+LINK_RE = re.compile(r"\[(.+?)\]\((.+?)\)")
+
+
+class DocumentBuilder(ABC):
```

Comment:
Solid, my only question here is are we trying to be too complete with what we can process? Would users get as much value being able to only have plain text?

Reply:
Yeah I think we could open ourselves up to a whole world of pain 🤣, lets see what gets picked up from use and then expand from there?

Reply:
OK, makes sense.